From fe8fbc92920d4fe6da6a4d49fac366e45db23ad7 Mon Sep 17 00:00:00 2001
From: Dave Young <dyoung@redhat.com>
Date: Thu, 22 May 2014 02:43:07 -0400
Subject: [firmware] efi: Check krealloc return value

Message-id: <1400726607-13717-16-git-send-email-dyoung@redhat.com>
Patchwork-id: 80116
O-Subject: [RHEL7.1 PATCH 15/35] efi: Check krealloc return value
Bugzilla: 1080109
RH-Acked-by: Chao WANG <chaowang@redhat.com>
RH-Acked-by: Vivek Goyal <vgoyal@redhat.com>

Resolves: bz1080109

Brew build:
https://brewweb.devel.redhat.com/taskinfo?taskID=7480539

Test: Tested kexec/kdump on HP workstation and qemu.

Upstream commit:
commit ee41143027706d9f342dfe05487a00b20887fde7
Author: Borislav Petkov <bp@suse.de>
Date:   Thu Oct 31 17:25:09 2013 +0100

    x86/efi: Check krealloc return value

    Check it just in case. We might just as well panic there because runtime
    won't be functioning anyway.

    Signed-off-by: Borislav Petkov <bp@suse.de>
    Signed-off-by: Matt Fleming <matt.fleming@intel.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/platform/efi/efi.c b/arch/x86/platform/efi/efi.c
index 3f4646d..f8ec4da 100644
--- a/arch/x86/platform/efi/efi.c
+++ b/arch/x86/platform/efi/efi.c
@@ -863,6 +863,9 @@ void __init efi_enter_virtual_mode(void)
   new_memmap = krealloc(new_memmap,
           (count + 1) * memmap.desc_size,
           GFP_KERNEL);
+  if (!new_memmap)
+   goto err_out;
+
   memcpy(new_memmap + (count * memmap.desc_size), md,
          memmap.desc_size);
   count++;
@@ -917,6 +920,11 @@ void __init efi_enter_virtual_mode(void)
     EFI_VARIABLE_BOOTSERVICE_ACCESS |
     EFI_VARIABLE_RUNTIME_ACCESS,
     0, NULL);
+
+ return;
+
+ err_out:
+ pr_err("Error reallocating memory, EFI runtime non-functional!\n");
 }
 
 /*
-- 
1.7.1